Update WeakRefs feature to match upcoming web.dev LIVE talk (#421)
v8/v8.dev: src/features/weak-references.md

---
title: 'Weak references and finalizers'

> author: 'Sathya Gunasekaran ([@_gsathya](https://twitter.com/_gsathya)), Mathias Bynens ([@mathias](https://twitter.com/mathias)), Shu-yu Guo ([@_shu](https://twitter.com/_shu)), and Leszek Swirski ([@leszekswirski](https://twitter.com/leszekswirski))

Reviewer: @syg This is probably the reason - someone removed the closing quote :)

Author: That someone would be this doofus. :)

  }

  compute(n) {
    // Compute the simple moving average for the last n events.
    // ...
  }
}
```

It is used by a `MovingAvgComponent` class that lets you control when to start and stop monitoring the simple moving average of the latency.

```js
class MovingAvgComponent {
  constructor(socket) {
    this.socket = socket;
  }

  start() {
    this.movingAvg = new MovingAvg(this.socket);
  }

  stop() {
    // Allow the garbage collector to reclaim memory.
    this.movingAvg = null;
  }

  render() {
    // Do rendering.
    // ...
  }
}
```

We know that keeping all the server messages inside an instance `MovingAvg` uses a lot of memory, so we take care to null out `this.movingAvg` when monitoring is stopped to let the garbage collector reclaim memory.

However, after checking in the memory panel in DevTools, we found out that memory was not being reclaimed at all! The seasoned web developer may have already spotted the bug: event listeners are strong references and must be explicitly removed. Thus, the listener in `MovingAvg` instances, by referencing `this`, keeps the whole instance alive as long as the event listener isn't removed.

> Until now, the solution is to manually unregister the event listener via `dispose` method.

Reviewer 2: ```suggestion
Until now, the solution is to manually unregister the event listener via the `dispose` method.
```

Reviewer 3: I would say "via **a** dispose method"

Reviewer 2: Right, that's better indeed! Edited the suggestion

Author: Done.

```js
class MovingAvg {
  constructor(socket) {
    this.events = [];
    this.socket = socket;
    this.listener = (ev) => { this.events.push(ev); };
    socket.addEventListener('message', this.listener);
  }

  dispose() {
    this.socket.removeEventListener('message', this.listener);
  }

>   // ...

Reviewer 2: Nit:

```suggestion
  // …
```

Author: Done.

}
```

> The downside to this approach is that it is manual memory management. `MovingAvgComponent`, and all other users of the `MovingAvg` class, must remember to call `dispose` or suffer memory leaks. Because the application behavior doesn't depend on the event listener of this diagnostic class, and that the listener is expensive in terms of memory use but not in computation, we really want the listener's lifetime to be logically tied to the `MovingAvg` instance. If its lifetime were tied to the `MovingAvg` instance, then `MovingAvg` could be used like any other JavaScript object whose memory is automatically reclaimed by the garbage collector.

Reviewer 2: “because […] and that the listener” sounds weird to my ESL ears. Maybe something like this?

```suggestion
The downside to this approach is that it is manual memory management. `MovingAvgComponent`, and all other users of the `MovingAvg` class, must remember to call `dispose` or suffer memory leaks. Because the application behavior doesn’t depend on the event listener of this diagnostic class, and the listener is expensive in terms of memory use but not in computation, we really want the listener's lifetime to be logically tied to the `MovingAvg` instance. If its lifetime were tied to the `MovingAvg` instance, then `MovingAvg` could be used like any other JavaScript object whose memory is automatically reclaimed by the garbage collector.
```

Feel free to reject if this is actually correct, of course.

Reviewer 3: I would split the sentence differently, e.g.

> The application behavior doesn't depend on the event listener of this diagnostic class, and the listener is expensive in terms of memory use but not in computation. What we really want is for the listener's lifetime to be logically tied to the `MovingAvg` instance, so that `MovingAvg` could be used like any other JavaScript object whose memory is automatically reclaimed by the garbage collector.

Author: Done.

`WeakRef`s make it possible to solve the dilemma by creating a _weak reference_ to the actual event listener, and then wrapping that `WeakRef` in an outer event listener. This way, the garbage collector can clean up the actual event listener and the memory that it holds alive, like the `events` array.

```js
class MovingAvg {
  constructor(socket) {
    this.events = [];

>     const listener = (ev) => { this.events.push(ev); }

Reviewer 2: ```suggestion
    const listener = (ev) => { this.events.push(ev); };
```

Reviewer 2: For reference, we try to follow [the Google JS style guide](https://google.github.io/styleguide/jsguide.html) for any public-facing examples in blog posts and slides, which means: use semicolons, use single quotes instead of double quotes for string literals, prefer `const` over `let`, etc.

Reviewer: I'd remove the braces altogether since the return value doesn't matter:

```suggestion
    const listener = (ev) => this.events.push(ev);
```

Reviewer 2: Interestingly, for the exact same reason I would keep the braces :) (Only `return` something non-`undefined` when you need it.)

Reviewer: Heh, usually when the caller expects `undefined`, I just return any value.

I know what the ultimate solution would be 😀
```suggestion
    const listener = (ev) => void this.events.push(ev);
```

Author: Added semicolon.

>     const weakRef = new WeakRef(listener);

Author: Ah, doh, this is difficult to get right. I wonder how to keep it alive without making it uncollectible...

>     const weakRef = new WeakRef(listener);

Author: FWIW since it's part of the WeakRef, at least it'll be guaranteed to stay alive for *1* turn. 😅

>     const weakRef = new WeakRef(listener);

Reviewer 4: I think you need to `WeakRef(this)`:

```js
class MovingAvg {
  constructor(socket) {
    this.events = [];
    this.listener = (ev) => { this.events.push(ev); }

    const weakRef = new WeakRef(this);
    const listenerWrapper = (ev) => {
      const deref = weakRef.deref();
      if (deref !== undefined) deref.listener(ev);
    });
    socket.addEventListener("message", listenerWrapper);
  }
}

>     const weakRef = new WeakRef(listener);

Author: That's trickier to reason through. I plan to update with the following, which is easier to understand and side-steps the non-split function context subtlety:

```js
function makeWeakListener(listener) {
  const weakRef = new WeakRef(listener);
  const wrapper = (ev) => { weakRef.deref()?.(ev); };
  return wrapper;
}

class MovingAvg {
  constructor(socket) {
    this.events = [];
    this.listener = (ev) => { this.events.push(ev); };
    socket.addEventListener("message", makeWeakListener(this.listener));
  }
}
```

As long as you make the wrapper closure outside of the lexical scope of the constructor itself, you can sidestep the trickiness of accidentally entraining variables in the function context.

>     const weakRef = new WeakRef(listener);

Author: To further clarify why `WeakRef(this)` doesn't work: once `FinalizationRegistry` comes into the picture, we need to pass the wrapper listener to it as a holding value so the finalizer can remove the event listener. That wrapper is strongly referenced as the holdings of the FR. If that wrapper is created inside the lexical scope of the constructor, it'll end up entraining all the other closed over variables. Since `this.listener` is an arrow and closes over `this`, the wrapper thus keeps the `MovingAvg` instance alive. It gets worse: since that `MovingAvg` instance is alive and that instance has a `this.listener` property, that arrow is also kept alive by the FR.

Unfortunately, function contexts is a pretty big footgun that devs will need to understand to use WeakRefs and FRs correctly with functions.

    const listenerWrapper = (ev) => {

>       const deref = weakRef.deref();

Reviewer 3: nit: call this "listener", and we could even call the weakRef "weakListener"

Author: I refactored the wrapping and making of the new WeakRef into a separate function this hopefully is less confusing?

>       if (deref !== undefined) deref(ev);

Reviewer 3: Explicit for ease-of-reading? I did kinda like the optional chain function call for this.

Author: I was trying to stay period accurate to the time the original post was written.

Reviewer: Since it's going to be updated anyway, I'd change to the new syntax. Just make sure to add `updated:` date to frontmatter, which will be rendered once #420 is merged.

    });

>     socket.addEventListener("message", listenerWrapper);

Reviewer 2: Coding style consistency nit:

```suggestion
    socket.addEventListener('message', listenerWrapper);
```

Author: Changed to "" everywhere.

Author: Oops, changing it to '' everywhere.

>     socket.addEventListener("message", listenerWrapper);

Reviewer 5: Doesn't this still leak the listenerWrapper? The socket doesn't know that the weakref died, and doesn't remove listenerWrapper.

In general, aren't the weak refs "off by one layer" here? The eventlistener list should be a list of weakrefs, so that we can clean up the eventlisteners which are no longer relevant.

Or did you mean the cleanup would be done by listenerwrapper when it gets an event and it notices its weakref has expeired?

Reviewer 3: It does indeed leak the listenerWrapper, which is the point of the subsequent finalization registry note.

Reviewer 5: Ah, right, I was confused by the "No more dispose!" below which I took to mean "dispose is not needed now, yay". Which is confusing, like, it's not needed in the first place if leaking is OK...

Reviewer 5: (It might be less confusing to still have the dispose here because we still need to remove the event listener, like in the previous version. But I realize it is probably way too late to fix this, so pls feel free to ignore this comment.)

Author: Removed the comment that said "No more dispose!"

  }

>   // No more dispose!

Reviewer 2: ```suggestion
  // No more `dispose`!
```

Author: Done.

:::note
**Note:** The finalization callback does not run immediately after garbage-collecting the event listener. It either runs at some point in the future, or not at all — the spec doesn’t guarantee that it runs! Keep this in mind when writing code.
:::

We can register a callback with a `FinalizationRegistry` to remove `listenerWrapper` from the socket when the inner event listener is garbage-collected. Our final implementation looks like this:

```js
const gListenersRegistry = new FinalizationRegistry(
  ({ socket, listenerWrapper }) => {
    socket.removeEventListener(socket, listenerWrapper); // 5
  });

class MovingAvg {
  constructor(socket) {
    this.events = [];
    const listener = (ev) => { this.events.push(ev); }
    const weakRef = new WeakRef(listener); // 1
    const listenerWrapper = (ev) => {
      const deref = weakRef.deref(); // 2
      if (deref !== undefined) deref(ev); // 3
    });

>     socket.addEventListener("message", listenerWrapper);

Reviewer 2: ```suggestion
    socket.addEventListener('message', listenerWrapper);
```

Author: Done.

>     gListenersRegistry.register(listener,

Reviewer: Nitpick: such split, while common in C / C++, looks a bit odd in typical JS. I'd either put it on the same line at the risk of running over max-length

```js
gListenersRegistry.register(listener, { socket, listenerWrapper });
```

or use block-style formatting

```js
gListenersRegistry.register(listener, {
  socket,
  listenerWrapper
});
```

Author: Made one line.

                                { socket, listenerWrapper }); // 4
  }
}
```

> We wrap the event listener that does the work in a `WeakRef` to make garbage-collectible, and to not leak its reference to the `MovingAvg` instance via `this` (1). We make a wrapper that `deref` the `WeakRef` to check if it is still alive (2), then call it if so (3). After registering the listener wrapper on the socket, we also register the inner listener on the `FinalizationRegistry`, passing a _holding value_ `{ socket, wrapper }` to the registration (4). Sometime after the inner listener garbage-collected, the finalizer may run, with the holding value passed to it. Inside the finalizer, we unregister the listener wrapper as well, making all memory associated with a `MovingAvg` instance garbage-collectible (5).

Reviewer 2: ```suggestion
We wrap the event listener that does the work in a `WeakRef` to make garbage-collectible, and to not leak its reference to the `MovingAvg` instance via `this` (1). We make a wrapper that `deref` the `WeakRef` to check if it is still alive (2), then call it if so (3). After registering the listener wrapper on the socket, we also register the inner listener on the `FinalizationRegistry`, passing a _holding value_ `{ socket, wrapper }` to the registration (4). Sometime after the inner listener gets garbage-collected, the finalizer may run, with the holding value passed to it. Inside the finalizer, we unregister the listener wrapper as well, making all memory associated with a `MovingAvg` instance garbage-collectible (5).
```

Author: Done.

With all this, our original implementation of `MovingAvgComponent` neither leaks memory nor requires any manual disposal.

## Don’t overdo it { #progressive-enhancement }

After hearing about these new capabilities, it might be tempting to `WeakRef` All The Things™. However, that’s probably not a good idea. Some things are explicitly _not_ good use cases for `WeakRef`s and finalizers.

> In general, avoid writing code that depends on the garbage collector cleaning up a `WeakRef` or calling a finalizer at any predictable time — [it can’t be done](https://github.com/tc39/proposal-weakrefs#a-note-of-caution)! Moreover, whether an object is garbage-collectible at all may depend on implementation details, such as the representation of closures, that are both subtle and may differ between different versions of V8.

Reviewer 2: Let’s not just talk about V8 here but also mention different JavaScript engines. Maybe something like:

```suggestion
In general, avoid writing code that depends on the garbage collector cleaning up a `WeakRef` or calling a finalizer at any predictable time — [it can’t be done](https://github.com/tc39/proposal-weakrefs#a-note-of-caution)! Moreover, whether an object is garbage-collectible at all may depend on implementation details, such as the representation of closures, that are both subtle and may differ between different versions of V8 and across JavaScript engines.
```

Author: Adopted.

```js
class MovingAvg {
  constructor(socket) {
    this.events = [];
    this.socket = socket;
    this.listener = (ev) => { this.events.push(ev); };
    socket.addEventListener('message', this.listener);
  }

  compute(n) {
    // Compute the simple moving average for the last n events.
    // ...
  }
}
```

It is used by a `MovingAvgComponent` class that lets you control when to start and stop monitoring the simple moving average of the latency.

```js
class MovingAvgComponent {
  constructor(socket) {
    this.socket = socket;
  }

  start() {
    this.movingAvg = new MovingAvg(this.socket);
  }

  stop() {
    // Allow the garbage collector to reclaim memory.
    this.movingAvg = null;
  }

  render() {
    // Do rendering.
    // ...
  }
}
```

We know that keeping all the server messages inside an instance `MovingAvg` uses a lot of memory, so we take care to null out `this.movingAvg` when monitoring is stopped to let the garbage collector reclaim memory.

> However, after checking in the memory panel in DevTools, we found out that memory was not being reclaimed at all! The seasoned web developer may have already spotted the bug: event listeners are strong references and must be explicitly removed. Thus, the listener in `MovingAvg` instances, by referencing `this`, keeps the whole instance alive as long as the event listener isn't removed.

Reviewer 3: This explanation might be better with explicit diagrams, e.g. https://drive.google.com/file/d/1Xca-f4PoieduOjoDMiOB2or16cUbI3u0/view?usp=sharing

Author: Oh wow, thanks for the diagrams!

Until now, the solution is to manually unregister the event listener via a `dispose` method.

```js
class MovingAvg {
  constructor(socket) {
    this.events = [];
    this.socket = socket;
    this.listener = (ev) => { this.events.push(ev); };
    socket.addEventListener('message', this.listener);
  }

  dispose() {
    this.socket.removeEventListener('message', this.listener);
  }

  // …
}
```

> The downside to this approach is that it is manual memory management. `MovingAvgComponent`, and all other users of the `MovingAvg` class, must remember to call `dispose` or suffer memory leaks. The application behavior doesn't depend on the event listener of this diagnostic class, and the listener is expensive in terms of memory use but not in computation. What we really want is for the listener's lifetime to be logically tied to the `MovingAvg` instance, so that `MovingAvg` could be used like any other JavaScript object whose memory is automatically reclaimed by the garbage collector.

Reviewer 3: Another downside I just thought of -- if MovingAvgComponent is collected before stopping, then there's no one that could call dispose.

Author: I'll add this sentence: What's worse, manual memory management is cascading: users of `MovingAvgComponent` must remember to call `stop` or suffer memory leaks, so on and so forth.

`WeakRef`s make it possible to solve the dilemma by creating a _weak reference_ to the actual event listener, and then wrapping that `WeakRef` in an outer event listener. This way, the garbage collector can clean up the actual event listener and the memory that it holds alive, like the `MovingAvg` instance and its `events` array.

```js
function addWeakListener(socket, listener) {
  const weakRef = new WeakRef(listener);
  const wrapper = (ev) => { weakRef.deref()?.(ev); };
  socket.addEventListener('message', wrapper);
}

class MovingAvg {
  constructor(socket) {
    this.events = [];
    this.listener = (ev) => { this.events.push(ev); };
    addWeakListener(socket, this.listener);
  }
}
```

:::note

> **Note:** `WeakRef`s to functions must be treated with caution. JavaScript functions are [closures](https://en.wikipedia.org/wiki/Closure_(computer_programming)) and strongly reference the outer environments which contain the values of free variables referenced inside the functions. These outer environments may contain variables that _other_ closures reference as well. That is, when dealing with closures, their memory is often strongly referenced by other closures in subtle ways. Keep this in mind when writing code.

Reviewer 3: Say explicitly that this is why we have a separate `addWeakListener` function, and that it's explicitly not local to the MovingAvg constructor to not be part of its lexical scope.

Author: Will add.

:::

We first make the event listener and assign it to `this.listener`, so that it is strongly referenced by the `MovingAvg` instance. In other words, as long as the `MovingAvg` instance is alive, so is the event listener.

Then, in `addWeakListener`, we create a `WeakRef` whose _target_ is the actual event listener. Inside `wrapper`, we `deref` it. Because `WeakRef`s do not prevent garbage collection of their targets if the targets do not have other strong references, we must manually dereference them to get the target. If the target has been garbage-collected in the meantime, `deref` returns `undefined`. Otherwise, the original target is returned, which is the `listener` function we then call using [optional chaining](/features/optional-chaining).

Since the event listener is wrapped in a `WeakRef`, the _only_ strong reference to it is the `listener` property on the `MovingAvg` instance. That is, we've successfully tied the lifetime of the event listener to the lifetime of the `MovingAvg` instance.

> But there's still a problem here: we've added a level of indirection to `listener` by wrapping it a `WeakRef`, but the wrapper in `addWeakListener` is still leaking for the same reason that `listener` was leaking originally. The solution to this is the companion feature to `WeakRef`, `FinalizationRegistry`. With the new `FinalizationRegistry` API, we can register a callback to run when the garbage collector zaps a register object. Such callbacks are known as _finalizers_.

Reviewer 3: Is it worth noting explicitly that the leak isn't as bad as before, because we're only leaking the one closure, not the whole MovingAvg class?

Author: Will clarify.

:::note
**Note:** The finalization callback does not run immediately after garbage-collecting the event listener. It either runs at some point in the future, or not at all — the spec doesn’t guarantee that it runs! Keep this in mind when writing code.
:::

> We can register a callback with a `FinalizationRegistry` to remove `listenerWrapper` from the socket when the inner event listener is garbage-collected. Our final implementation looks like this:

Reviewer 3: nit: wrapper

```js

> const gListenersRegistry = new FinalizationRegistry(

Reviewer 3: Is it worth mentioning that gListenersRegistry here is global, because the lifetime of the registry is independent of the lifetime of the weakrefs?

Author: Yeah, good call. Seems worth a note.

>   ({ socket, wrapper }) => { socket.removeEventListener(socket, wrapper); }); // 6

Reviewer 5: Should the first param to removeEventListener be "message" instead of socket?

Author: Ah oops, nice catch. Will fix.
